kubelet: change node-lease-renew-interval to 0.25 of lease-renew-duration

gaorong opened the pull request:
**What type of PR is this?**
/kind feature

Ref: https://github.com/kubernetes/enhancements/issues/589

**What this PR does / why we need it**:
change node-lease-renew-interval to 0.25 of lease-renew-duration instead of hardcoded `10s` 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
kubelet: change node-lease-renew-interval to 0.25 of lease-renew-duration
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

@wojtek-t @liggitt @wangzhen127 @mtaufen

mattjmcnaughton reviewed (commented):
Thanks for your Pr :)

Have you chatted with anyone about whether we consider this change an API change? My guess is yes, in which case we will want to follow the API review process, which is documented here: https://github.com/kubernetes/community/blob/master/sig-architecture/api-review-process.md.

wojtek-t commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -170,6 +170,8 @@ type KubeletConfiguration struct {
 	// frequency and post node status immediately if any change is detected. It is
 	// only used when node lease feature is enabled.
 	NodeStatusReportFrequency metav1.Duration
+	// nodeLeaseRenewFrequency is the frequency that kubelet renews its corresponding Lease.
```
when NodeLease feature is enabled:
https://github.com/kubernetes/kubernetes/blob/323356fda06cbac0b74c35f522f17c3bb9d580ca/pkg/features/kube_features.go#L544

wojtek-t commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -328,15 +328,23 @@ type KubeletConfiguration struct {
+	// NodeLifecycleController.
+	// Default: "10s"
+	// +optional
+	NodeLeaseRenewFrequency metav1.Duration `json:"nodeLeaseRenewFrequency,omitempty"`
```
Yes - this formally requires api review.

mtaufen commented:
@wojtek-t @bgrant0607 Didn't (a long time ago) we discuss that it would be better to determine a reasonable fraction of the lease duration to renew (say 1/2, or 3/4, or something) that works well in virtually all cases, probably based on operational data, and just derive the period between renewals from that?

gaorong commented:
As a reminder, there are some discussions about this at first PR regarding node lease(see [here](https://github.com/kubernetes/kubernetes/pull/66257#issuecomment-414853917)).

IMHO, cluste- config is a great suggestion in the long run, but may need a long time to come true and then be production available, we should not take that as a blocker of other functionality (such as node lease).

wojtek-t commented:
> @wojtek-t @bgrant0607 Didn't (a long time ago) we discuss that it would be better to determine a reasonable fraction of the lease duration to renew (say 1/2, or 3/4, or something) that works well in virtually all cases, probably based on operational data, and just derive the period between renewals from that?

I can't remember it (though it doesn't mean we didn't).
I missed the fact the NodeLeaseDuration is already exposed via this config (I though it is hardcoded similarly to renew-interval).
In that case, that sounds like a reasonable option to me. I think 1/2 or way to much (one missed update would mean that we can loose the lease). But 1/4th I think it would actually make sense.

gaorong commented:
> determine a reasonable fraction of the lease duration to renew

what about `node-monitor-grace-period`? 
For now, the NodeLifecycleController in controller-manager marks node as "Ready: Unknown"  if it does not get heartbeat message at every  "node-monitor-grace-period" cycle, and it does not care about lease-duration at all.

If a user has a big value of `lease duration`(as we allow user specify this value), let's say 4min, then the lease interval is 1min (1/4 * 4min), if controller-manager has a `node-monitor-grace-period` whose value is 40s, In this case, the node will always be ready-unknown state.

If we want to determine lease interval based on lease duration, we must make NodeLifecycleController follow lease-duration and ignore node-monitor-grace-period value when node lease feature is enabled. I'm not sure if this is wanted. 

aside: IMHO, before we want to GA node lease feature, we may want to implement node-duration at NodeLifecycleController side, for now, `leaseDuriationSeconds` is currently only propagated as part of Lease object and not really used in any meaningful way. this may be a good chance to make a change.

dims commented:
/ok-to-test
/uncc

wojtek-t commented:
> If a user has a big value of lease duration(as we allow user specify this value), let's say 4min, then the lease interval is 1min (1/4 * 4min), if controller-manager has a node-monitor-grace-period whose value is 40s, In this case, the node will always be ready-unknown state.

Users can't change one of those values (lease duration) without thinking and most probably changing the other one (node-monitor-grace-period).
Exactly the same problem exists without node-lease - you can't increase node-update-frequency without touching node-mointor-grace-period.
With node-lease, we are just saying that dependencies are a little bit more complex.

> If we want to determine lease interval based on lease duration, we must make NodeLifecycleController follow lease-duration and ignore node-monitor-grace-period value when node lease feature is enabled. I'm not sure if this is wanted.

As mentioned above - I don't agree that there is such implication.

> aside: IMHO, before we want to GA node lease feature, we may want to implement node-duration at NodeLifecycleController side, for now, leaseDuriationSeconds is currently only propagated as part of Lease object and not really used in any meaningful way. this may be a good chance to make a change.

I disagree it's a release blocker for GA - this is a separate feature.

gaorong commented:
yes, I agree users always consider those two parameters together and I totally understand you guys well-intention (which can free users from complicated configuration)

In most cases, users would firstly generate a `node-monitor-grace-period` based on how long time their workload can tolerate if some hosts lost heartbeat. then depending on  `node-monitor-grace-period`, they generate another parameter: `node-update-frequency` (or `node-lease-renew-interval` in node lease situation). 

If we want to determine lease interval based on lease duration,  we must tune `node-lease-renew-interval` via calculating and changing `node-lease-duration` (with the precondition they know the fraction is 1/4 or sth, this is also hard for most user unless we have a good document), which seems a little bit weird especially node-lease-duration is not used by controller-manager for now.

wojtek-t commented:
> If we want to determine lease interval based on lease duration, we must tune node-lease-renew-interval via calculating and changing node-lease-duration (with the precondition they know the fraction is 1/4 or sth, this is also hard for most user unless we have a good document), which seems a little bit weird especially node-lease-duration is not used by controller-manager for now.

I guess what I wanted to say is:
"with node lease the situation is not much worse than it is without leases".

Without node lease:
- you determine "node-monitor-grace-period"
- based on that you need to find out what is the reasonable value for "node-update-frequency" [we're generally saying 1/4th of node-monitor-grace-period]

With node lease (with what @mtaufen suggest and I believe is reasonable)
- you determine "node-monitor-grace-period"
- based on that you set "node-lease-duration" to the exactly the same value and we will automatically set "renew-interval" to 1/4th of it

I would say that this would be even simpler, because we will say that we suggest using exactly the same value of those flags so users don't need to think how to compute the latter.

Thoughts?

gaorong commented:
> * based on that you set "node-lease-duration" to the exactly the same value and we will automatically set "renew-interval" to 1/4th of it

this seems more reasonable. 

Another thing to be considered:
In a long term, we may want to deprecate `node-monitor-grace-period` and let lease duration be the only basis of "amount of time which we mark node unhealthy", as those two flags truly have the same functionality and may conflict with each other. this action can also free user from the burden of "setting node-lease-duration to the exactly the same value of node-monitor-grace-period".

The reason why I prefer lease duration to `node-monitor-grace-period` is: 
each node can have different lease duration and which leave more free to user, and this may be useful for heterogeneous cluster which nodes spread different AZ and have different network latency.

If this makes sense to you, I'm going to push this forward. (open a PR to update the node lease KEP and let more people involved in to discuss)

wojtek-t commented:
> If this makes sense to you, I'm going to push this forward. (open a PR to update the node lease KEP and let more people involved in to discuss)

I think it's worth starting discussion about.
Though it shouldn't be modification of existing KEP - it should be a new KEP.


BTW - based on the whole above discussion - I guess we want to change the logic to set renew-interval based on node-lease-duration - would you be able to change the PR to do that (or open a new one)?

gaorong committed:
change node-lease-renew-interval to 0.25 of renew-duration

0.25 is a dedicated value to align before default value
of renew-interval but get more heuristic interval

wojtek-t commented on test/conformance/testdata/conformance.txt:
```diff
@@ -15,6 +15,7 @@ test/e2e/apimachinery/watch.go: "should be able to start watching from a specifi
 test/e2e/apimachinery/watch.go: "should be able to restart watching from the last resource version observed by the previous watch"
 test/e2e/apimachinery/watch.go: "should observe an object deletion if it stops meeting the requirements of the selector"
 test/e2e/apimachinery/watch.go: "should receive events on concurrent watches in same order"
+test/e2e/apimachinery/table_conversion.go: "should return a 406 for a backend which does not implement metadata"
```
Please remove

gaorong replied:
oh, my mistake, still work in progress at that time
now it's ready to review, please take a look.

spiffxp commented:
/remove-area conformance

gaorong commented:
/retest

gaorong commented:
the CI is green now.

wojtek-t commented on pkg/kubelet/nodelease/controller.go:
```diff
@@ -60,27 +60,18 @@ type controller struct {
-		renewInterval = nodeStatusUpdateFrequency
-	}
-
+	renewIntervalSeconds := int64(renewIntervalFraction * float64(leaseDurationSeconds))
```
It's not what we want - if you set say leaseDurationSecond=3, then this one will be 0 (which definitely isn't what we want).
Should be something like:
```
leaseDuration := time.Duration(leaseDurationSeconds) * time.Second
renewInterval := time.Duration(float64(leaseDuration) * renewIntervalFraction)
``

gaorong replied:
done, thanks

wojtek-t commented:
@gaorong - also please update PR title and description

wojtek-t commented:
/lgtm
/approve
